### PR TITLE
Add timeout in sending message rtc offer creation to make firefox happy

### PR DESCRIPTION
**Pull request by willscott**
Fixes 1 of the 3 failing tests covered by https://github.com/freedomjs/freedom-for-firefox/issues/29

### CODE REVIEW

**Comment by willscott** [comment]
This works here, but similar fix doesn't seem sufficient for other 2 tests.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 1]
```diff
@@ -40,6 +40,15 @@ module.exports = function (pc, dc, setup) {
       });
       alice.createOffer().then(function (offer) {
         return alice.setLocalDescription(offer).then(function () {return offer; });
+      }).then(function() {
+        // Delay 100ms for Firefox to get its act together.
+        return new Promise(function(resolve) {
+          setTimeout(function() {
+            alice.createOffer().then(function (offer) {
```
Comment (ryscheng) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: What is the behavior of calling createOffer twice? @bemasc

Comment (bemasc) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: `createOffer` is a const-method; it doesn't mutate the RTCPeerConnection.  However, it's not supposed to be necessary to call it multiple times (unless `.onnegotiationneeded` fires).  The only difference we've seen so far is the inclusion of more local candidates ... but those candidates should also be provided via `.onicecandidate`.

It should be safe, but it shouldn't be necessary.

**Comment by ryscheng** [comment]
Cool. @bemasc Does this look good to you then? 
:+1:  From me

**Comment by willscott** [comment]
I guess it sounds like a longer term solution is that our tests probably
need to look at o'n negotiation needed' to work in Firefox reliably?

**Comment by willscott** [comment]
Restructured this to a very conservative code path:

* generate alice's offer.
* wait until all ice candidates are produced
* set bob's remote description to offer
* give bob ALice's ice candidtaes
* get the answer and give back to alice.

With this, it seems to pass reliably on both firefox 34/mac and 35/linux.

Will extend this to the other two tests in this file.

**Comment by willscott** [comment]
All 3 tests written in the new style.
The stats test doesn't pass, because the stats object returned from the getStats call shows on the console as
```
Object
  - prototype Object
```
and has no iterable keys, as the test expects. @bemasc any idea what should be coming out of the getStats call, and whether this is a webrtc-adapter issue or firefox not being in the expected state?

From here, the next step is to merge the refactor to this test from #203, to get a test for binary data transfer.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 2]
```diff
@@ -15,11 +15,9 @@ module.exports = function (pc, dc, setup) {
   });
 
   it("Sends messages across data channels", function (done) {
-    var alice, bob, aliceChannel, bobchannel;
+    var alice, bob, aliceChannel, bobchannel,
```
Comment (bemasc) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: capitalization consistency: aliceChannel vs bobchannel

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 3]
```diff
@@ -33,23 +31,35 @@ module.exports = function (pc, dc, setup) {
...
-      }).then(function () {
-        return bob.createAnswer();
-      }).then(function (answer) {
-        return bob.setLocalDescription(answer).then(function () {return answer; });
-      }).then(function (answer) {
-        return alice.setRemoteDescription(answer);
-      }, function (err) {
-        console.error('RTC failed: ',err);
+        alice.setLocalDescription(offer);
+        aliceCandidates.push(offer);
```
Comment (bemasc) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: This seems pretty confusing to me.  I'd rather have a separate `aliceOffer` variable.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 4]
```diff
@@ -33,23 +31,35 @@ module.exports = function (pc, dc, setup) {
...
+        bob.setRemoteDescription(aliceCandidates.shift()).then(function () {
+          var last;
+          while (aliceCandidates.length) {
+            last = bob.addIceCandidate(aliceCandidates.shift())
+          }
+          return last;
+        }).then(function () {
+          return bob.createAnswer();
+        }).then(function (answer) {
+          return bob.setLocalDescription(answer).then(function () {return answer; });
```
Comment (bemasc) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: There's no need to sequence `bob.setLocal` and `alice.setRemote`, so they can both be in one promise callback.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 5]
```diff
@@ -33,23 +31,35 @@ module.exports = function (pc, dc, setup) {
     bob.on('onicecandidate', function (msg) {
       msg.candidate && alice.addIceCandidate(msg.candidate);
     });
+    alice.on('onicecandidate', function (msg) {
+      if (!msg.candidate) {
```
Comment (bemasc) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: We should have a comment explaining what bug we think we are working around.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 6]
```diff
@@ -61,12 +71,16 @@ module.exports = function (pc, dc, setup) {
...
   it("Closes Cleanly", function (done) {
-    var alice, bob, aliceChannel, bobchannel;
+    var alice, bob, aliceChannel, bobchannel,
+      aliceCandidates = [];
     var onClose = jasmine.createSpy('onclose');
     onClose.and.callFake(function () {
       // TODO: Firefox doesn't yet close remote data channels.
-      if (onClose.calls.count() === 2 ||
-         navigator.userAgent.indexOf("Firefox") > 0) {
+      console.error('onclose');
```
Comment (bemasc) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: Left over, presumably.

Comment (willscott) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: done.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js (hunk 3 above)
Comment (willscott) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: done.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js (hunk 4 above)
Comment (willscott) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: i'm going to err on the side of caution and keep them in a known-working configuration.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js (hunk 2 above)
Comment (willscott) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: done.

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 7]
```diff
@@ -61,12 +73,15 @@ module.exports = function (pc, dc, setup) {
...
-    var alice, bob, aliceChannel, bobchannel;
+    var alice, bob, aliceChannel, bobChannel,
+      aliceCandidates = [],
+      aliceOffer;
     var onClose = jasmine.createSpy('onclose');
     onClose.and.callFake(function () {
       // TODO: Firefox doesn't yet close remote data channels.
-      if (onClose.calls.count() === 2 ||
-         navigator.userAgent.indexOf("Firefox") > 0) {
+      if (onClose.calls.count() === 2 || (typeof navigator !== 'undefined' &&
```
Comment (ryscheng) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: Weird line break and why compare >0?

File: spec/providers/coreIntegration/rtcpeerconnection.integration.src.js [hunk 8]
```diff
@@ -96,30 +129,18 @@ module.exports = function (pc, dc, setup) {
...
+        aliceOffer = offer;
       });
     }, function (err) {
       console.error('RTC failed: ',err);
     });
   });
 
-  it("getStats works", function (done) {
-    var alice, bob;
+  fit("getStats works", function (done) {
```
Comment (ryscheng) [inline spec/providers/coreIntegration/rtcpeerconnection.integration.src.js]: fit?

**Comment by ryscheng** [comment]
:+1: except for minor changes

**Comment by bemasc** [comment]
:+1:

**Pull request merged**